### PR TITLE
Backbone fixes

### PR DESCRIPTION
**Pull request by khasanovbi**
- Update Model methods definitions

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/jashkenas/backbone/blob/master/backbone.js>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/backbone/index.d.ts [hunk 1]
```diff
@@ -293,34 +293,32 @@ declare namespace Backbone {
...
-        clear(options?: Silenceable): any;
+        changedAttributes(attributes?: Partial<T>): Partial<T> | false;
+        clear(options?: Silenceable): this;
         clone(): Model;
-        destroy(options?: ModelDestroyOptions): any;
-        escape(attribute: string): string;
-        has(attribute: string): boolean;
-        hasChanged(attribute?: string): boolean;
+        destroy(options?: ModelDestroyOptions): JQueryXHR | false;
+        escape(attribute: keyof T & string): string;
```
Comment (jgonggrijp) [inline types/backbone/index.d.ts:300]: Why not just `keyof T`?

File: types/backbone/index.d.ts [hunk 2]
```diff
@@ -293,34 +293,32 @@ declare namespace Backbone {
...
-        save(attributes?: any, options?: ModelSaveOptions): any;
-        unset(attribute: string, options?: Silenceable): Model;
-        validate(attributes: any, options?: any): any;
-
-        private _validate(attributes: any, options: any): boolean;
+        previous<a extends keyof T & string>(attribute: a): T[a]| null | undefined;
+        previousAttributes(): Partial<T>;
+        save(attributes?: Partial<T> | null, options?: ModelSaveOptions): JQueryXHR;
+        unset(attribute: keyof T & string, options?: Silenceable): this;
+        validate(attributes: Partial<T>, options?: any): any;private _validate(attributes: any, options: any): boolean;
```
Comment (jgonggrijp) [inline types/backbone/index.d.ts:309]: This was probably an accidental deletion.

```suggestion
        validate(attributes: Partial<T>, options?: any): any;
        
        private _validate(attributes: any, options: any): boolean;
```

File: types/backbone/index.d.ts [hunk 3]
```diff
@@ -241,7 +241,7 @@ declare namespace Backbone {
         * For assigning an object hash, do it like this: this.defaults = <any>{ attribute: value, ... };
         * That works only if you set it in the constructor or the initialize method.
         **/
-        defaults(): ObjectHash;
+        defaults(): Partial<T>;
```
Comment (jgonggrijp) [inline types/backbone/index.d.ts:244]: In principle I'm in favor, but if you go this route, you should also update the properties on lines 228-236.

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
I definitely like these changes. Just some details that might need polishing. I'd also suggest adding a few tests to check that the optional improved attribute type checking actually works when people decide to set `T` to something other than `any`.

File: types/backbone/index.d.ts (hunk 1 above)
Comment (khasanovbi) [reply types/backbone/index.d.ts:300]: Because when `T = any`, `keyof T = any` too, but attr name should be always a string

**Review by jgonggrijp — APPROVED** [review]
I'm in favor as-is, but you might want to extend your polymorphic `this` return type fixes to methods of other classes. If I'm right, `View.remove()` currently returns `Backbone.View` while it should be `this`, and there are probably a few other methods with that problem.

In any case, thanks a lot for these fixes!

**Comment by khasanovbi** [comment]
> I'm in favor as-is, but you might want to extend your polymorphic `this` return type fixes to methods of other classes. If I'm right, `View.remove()` currently returns `Backbone.View` while it should be `this`, and there are probably a few other methods with that problem.
> 
> In any case, thanks a lot for these fixes!

I want to fix other places too, but that fixes break a lot of dependent libraries, so I think it easier to look at that fixes in separate PRs

**Comment by khasanovbi** [comment]
Ready to merge

**Pull request merged**